### PR TITLE
Add DenseNet models

### PR DESCRIPTION
**Pull request by taehoonlee**
This PR adds DenseNet121, DenseNet169, DenseNet201, and those test codes.

### CODE REVIEW

File: keras/applications/densenet.py [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+from ..layers import Input
+from ..layers import MaxPooling2D
+from ..layers import ZeroPadding2D
+from ..utils.data_utils import get_file
+from ..engine.topology import get_source_inputs
+from .imagenet_utils import decode_predictions
+from .imagenet_utils import _obtain_input_shape
+
+
+__url_prefix__ = 'https://github.com/taehoonlee/deep-learning-models/' \
```
Comment (fchollet) [inline keras/applications/densenet.py]: Prefer listing all different weight files path as ALL_CAPS constants instead (like we do in the other applications).

File: keras/applications/densenet.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+        img_input = Input(shape=input_shape)
+    else:
+        if not K.is_keras_tensor(input_tensor):
+            img_input = Input(tensor=input_tensor, shape=input_shape)
+        else:
+            img_input = input_tensor
+
+    bn_axis = 3 if K.image_data_format() == 'channels_last' else 1
+
+    x = ZeroPadding2D(padding=((3, 3), (3, 3)))(img_input)
```
Comment (fchollet) [inline keras/applications/densenet.py]: You don't need this, you can just use `padding='same'` in the layer below.

File: keras/applications/densenet.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+    x = Conv2D(64, 7, strides=2, use_bias=False, name='conv1/conv')(x)
+    x = BatchNormalization(axis=bn_axis, epsilon=1e-5,
+                           name='conv1/bn')(x)
+    x = Activation('relu', name='conv1/relu')(x)
+    x = ZeroPadding2D(padding=((1, 1), (1, 1)))(x)
```
Comment (fchollet) [inline keras/applications/densenet.py]: Likewise, you don't need this.

File: keras/applications/densenet.py [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+                         'or the path to the weights file to be loaded.')
+
+    if weights == 'imagenet' and include_top and classes != 1000:
+        raise ValueError('If using `weights` as imagenet with `include_top`'
+                         ' as true, `classes` should be 1000')
+
+    # Determine proper input shape
+    input_shape = _obtain_input_shape(input_shape,
+                                      default_size=224,
+                                      min_size=224,
```
Comment (fchollet) [inline keras/applications/densenet.py]: Are you sure that's the minimum input size one can use?

File: keras/applications/densenet.py [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+                input_shape=None,
+                pooling=None,
+                classes=1000):
+    return DenseNet([6, 12, 48, 32],
+                    include_top, weights,
+                    input_tensor, input_shape,
+                    pooling, classes)
+
+
+def preprocess_input(x):
```
Comment (fchollet) [inline keras/applications/densenet.py]: Is this a different preprocessing mode than what we already cover in `imagenet_utils`?

Note that `preprocess_input` should be implemented to work with both Numpy arrays and symbolic tensors. It should also be efficient (no memory copy).

File: keras/applications/densenet.py [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+        x: input tensor.
+        growth_rate: float, growth rate at dense layers.
+        scope: string, block label.
+
+    # Returns
+        output tensor for the block.
+    """
+    bn_axis = 3 if K.image_data_format() == 'channels_last' else 1
+    x1 = BatchNormalization(axis=bn_axis, epsilon=1e-5,
+                            name="%s/0/bn" % scope)(x)
```
Comment (fchollet) [inline keras/applications/densenet.py]: Use `'` as quote character

Don't use `/` in layer names as this is disallowed in TF (and will be overwritten with something else, maybe `_`)

File: keras/applications/densenet.py [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+    """Instantiates the DenseNet architecture.
+
+    Optionally loads weights pre-trained
+    on ImageNet. Note that when using TensorFlow,
+    for best performance you should set
+    `image_data_format='channels_last'` in your Keras config
+    at ~/.keras/keras.json.
+
+    The model and the weights are compatible with both
+    TensorFlow and Theano. The data format
```
Comment (fchollet) [inline keras/applications/densenet.py]: What about CNTK? Does it work? I'd think it would.

File: keras/applications/densenet.py [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+    x1 = BatchNormalization(axis=bn_axis, epsilon=1e-5,
+                            name="%s/1/bn" % scope)(x1)
+    x1 = Activation('relu', name="%s/1/relu" % scope)(x1)
+    x1 = Conv2D(growth_rate, 3, padding='same', use_bias=False,
+                name="%s/2/conv" % scope)(x1)
+    x = Concatenate(axis=bn_axis, name="%s/concat" % scope)([x, x1])
+    return x
+
+
+def DenseNet(blocks, include_top, weights, input_tensor, input_shape,
```
Comment (fchollet) [inline keras/applications/densenet.py]: Provide default values for these arguments

**Comment by ahundt** [comment]
Don't forget the discussion in https://github.com/keras-team/keras/pull/8133, in particular we have a fully tested and integrated version in keras-contrib which is the version I'd suggest merging if any. Furthermore, there is a performance problem with every tensorflow implementation of DenseNets, which is waiting on https://github.com/tensorflow/tensorflow/issues/12948.

**Comment by ahundt** [comment]
See https://github.com/farizrahman4u/keras-contrib/blob/master/keras_contrib/applications/densenet.py

Also including @titu1994

**Comment by ahundt** [comment]
Just looked at this PR's code. The performance problem due to quadratic memory growth discussed in tensorflow/tensorflow#12948 is present in this PR #8800.

**Comment by titu1994** [comment]
@ahundt I've had a look at the code earlier today, and I think the implementation here is fine as it is meant for inference alone. 

The Keras application version has no need to be adaptable to different DenseNet configs, nor does it need the regularizers for training and additional code blocks for segmentation and such. If we were to incorporate my version, after stripping down it's extra stuff, would boil down to this PR anyway. 

In addition, as a pure inference model, it doesn't specifically require the optimizations either thankfully. However, fine-tuning on the larger versions will be a significant challenge. 

The contrib version should remain, for the purpose of training these configs, or different configs and also support segmentation models.

**Comment by ahundt** [comment]
@titu1994 Sure that's fair.

If this is merged, might I suggest the code explicitly print a performance warning on initialization regarding tensorflow/tensorflow#12948?

File: keras/applications/densenet.py (hunk 1 above)
Comment (taehoonlee) [reply keras/applications/densenet.py]: Updated.

File: keras/applications/densenet.py (hunk 6 above)
Comment (taehoonlee) [reply keras/applications/densenet.py]: Updated.

File: keras/applications/densenet.py (hunk 8 above)
Comment (taehoonlee) [reply keras/applications/densenet.py]: Updated.

File: keras/applications/densenet.py (hunk 7 above)
Comment (taehoonlee) [reply keras/applications/densenet.py]: Yes, I confirmed that the codes work for all three backends. Docs have been also updated.

File: keras/applications/densenet.py (hunk 2 above)
Comment (taehoonlee) [reply keras/applications/densenet.py]: @fchollet, Theano and Tensorflow produced different results without `ZeroPadding2D`.
I tested the following 4 cases with the below codes:
- explicit `ZeroPadding2D` followed by `Conv2D(padding='valid')` under Theano
- `Conv2D(padding='same')` under Theano
- explicit `ZeroPadding2D` followed by `Conv2D(padding='valid')` under TensorFlow
- `Conv2D(padding='same')` under TensorFlow

As a result, only the last item shows a different output. In other words, Theano and TensorFlow can have the same outputs (currently) only if `ZeroPadding2D` is added explicitly.
```python
%env KERAS_BACKEND=tensorflow
import numpy as np
from keras.models import Model
from keras.layers import Input, Conv2D, ZeroPadding2D

inputs = Input(shape=(224, 224, 1))

x1 = ZeroPadding2D(padding=(3, 3))(inputs)
x1 = Conv2D(1, 7, strides=2, use_bias=False)(x1)
model1 = Model(inputs, x1)

x2 = Conv2D(1, 7, strides=2, use_bias=False, padding='same')(inputs)
model2 = Model(inputs, x2)

w = [np.arange(49).reshape((7, 7, 1, 1))]
if K.backend() == 'tensorflow':
    w = [np.flip(np.flip(w[0], 0), 1)]
model1.set_weights(w)
model2.set_weights(w)

x = np.ones((1, 224, 224, 1))
print(model1.predict(x)[0,:3,:3,0])
print(model2.predict(x)[0,:3,:3,0])
```

File: keras/applications/densenet.py (hunk 5 above)
Comment (taehoonlee) [reply keras/applications/densenet.py]: Yes, I borrowed the preprocessing method from \`https://github.com/liuzhuang13/DenseNet\` which is the same as \`https://github.com/facebook/fb.resnet.torch\`. I think that this function is hard to be integrated into \`imagenet_utils\` because it includes broadcasted division which is not implemented in the current version of the three backends.

Comment (taehoonlee) [reply keras/applications/densenet.py]: Yes, I borrowed the preprocessing method from `https://github.com/liuzhuang13/DenseNet` which is the same as `https://github.com/facebook/fb.resnet.torch`. I think that this function is hard to be integrated into `imagenet_utils` because it includes broadcasted division which is not implemented in the current version of the three backends.

**Comment by taehoonlee** [comment]
Thank you for pointing out an important issue, @ahundt! I would also like to revise the PR memory efficiently but can't really think of an idea now. In my opinion, the performance of the current version is not bad in terms of both memory and speed. The `DenseNet121` requires only ~400MB and runs as fast as an `ResNet50` under my tests.

**Comment by taehoonlee** [comment]
BTW, the travis error does not seem to have occurred in this PR.

File: docs/templates/applications.md [hunk 9]
```diff
@@ -604,6 +608,70 @@ These weights are released under [the Apache License](https://github.com/tensorf
...
+```
+
+Optionally loads weights pre-trained
+on ImageNet. Note that when using TensorFlow,
+for best performance you should set
+`image_data_format='channels_last'` in your Keras config
+at ~/.keras/keras.json.
+
+The model and the weights are compatible with both
+TensorFlow and Theano. The data format
```
Comment (fchollet) [inline docs/templates/applications.md]: Also CNTK, right?

File: keras/applications/densenet.py [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+
+DENSENET121_WEIGHT_PATH = 'https://github.com/taehoonlee/deep-learning-models/releases/download/densenet/densenet121_weights_tf_dim_ordering_tf_kernels.h5'
+DENSENET121_WEIGHT_PATH_NO_TOP = 'https://github.com/taehoonlee/deep-learning-models/releases/download/densenet/densenet121_weights_tf_dim_ordering_tf_kernels_notop.h5'
+DENSENET169_WEIGHT_PATH = 'https://github.com/taehoonlee/deep-learning-models/releases/download/densenet/densenet169_weights_tf_dim_ordering_tf_kernels.h5'
+DENSENET169_WEIGHT_PATH_NO_TOP = 'https://github.com/taehoonlee/deep-learning-models/releases/download/densenet/densenet169_weights_tf_dim_ordering_tf_kernels_notop.h5'
+DENSENET201_WEIGHT_PATH = 'https://github.com/taehoonlee/deep-learning-models/releases/download/densenet/densenet201_weights_tf_dim_ordering_tf_kernels.h5'
+DENSENET201_WEIGHT_PATH_NO_TOP = 'https://github.com/taehoonlee/deep-learning-models/releases/download/densenet/densenet201_weights_tf_dim_ordering_tf_kernels_notop.h5'
+
+
+def dense(x, blocks, scope):
```
Comment (fchollet) [inline keras/applications/densenet.py]: To avoid semantic collision with `Dense` (and `tf.layers.dense`) please rename this to e.g. `dense_blocks` or `dense_block_series`

File: keras/applications/densenet.py [hunk 11]
```diff
@@ -0,0 +1,341 @@
...
+    """A dense block.
+
+    # Arguments
+        x: input tensor.
+        blocks: integer, the number of building blocks.
+        scope: string, block label.
```
Comment (fchollet) [inline keras/applications/densenet.py]: All mentions of "scope" in this file should be renamed to `name`. Scope is TF terminology and does not make sense in this context (there are no variable scopes in Keras).

File: keras/applications/densenet.py [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
+
+    # Returns
+        output tensor for the block.
+    """
+    for i in range(blocks):
+        x = block(x, 32, scope=scope + '_block' + str(i + 1))
+    return x
+
+
+def transition(x, reduction, scope):
```
Comment (fchollet) [inline keras/applications/densenet.py]: Please rename to e.g. `transition_block`.

File: keras/applications/densenet.py [hunk 13]
```diff
@@ -0,0 +1,341 @@
...
+    x = BatchNormalization(axis=bn_axis, epsilon=1.001e-5,
+                           name=scope + '_bn')(x)
+    x = Activation('relu', name=scope + '_relu')(x)
+    x = Conv2D(int(x._keras_shape[bn_axis] * reduction), 1, use_bias=False,
+               name=scope + '_conv')(x)
+    x = AveragePooling2D(2, strides=2, name=scope + '_pool')(x)
+    return x
+
+
+def block(x, growth_rate, scope):
```
Comment (fchollet) [inline keras/applications/densenet.py]: Please rename to e.g. `dense_block`

File: keras/applications/densenet.py [hunk 14]
```diff
@@ -0,0 +1,341 @@
...
+    # Returns
+        Preprocessed array.
+    """
+    x /= 255.
+    x[:, :, :, 0] -= 0.485
+    x[:, :, :, 1] -= 0.456
+    x[:, :, :, 2] -= 0.406
+    x[:, :, :, 0] /= 0.229
+    x[:, :, :, 1] /= 0.224
+    x[:, :, :, 2] /= 0.225
```
Comment (fchollet) [inline keras/applications/densenet.py]: This preprocessing is only implemented for `channels_last` data format. Please add support for both.

**Comment by taehoonlee** [comment]
Thank you for the comments @fchollet! All the naming-related comments have been addressed.

**Comment by ahundt** [comment]
Would you mind enabling the preprocessing to work in the input tensor case as well, if it wouldn't be too much trouble?

**Comment by taehoonlee** [comment]
Summary of replies:
- The updates of docs, parameters, naming convention, and styles: completed.
- The addition of DenseNet records to the performance table: completed.
- The integration of the preprocessing into `imagenet_utils`: completed but needs to be reviewed.
- The removal of the explicit `ZeroPadding2D` (@fchollet): not for now, better be addressed in another PR. (In the current implementation, Theano and TensorFlow can have the same outputs when `strides=2` only if `ZeroPadding2D` is added explicitly.)
- The performance problem (@ahundt): not for now, better be addressed in another PR. (In my opinion, the performance of the current version is not bad in terms of both memory and speed. The `DenseNet121` requires only ~400MB and runs as fast as an `ResNet50` under my tests.)
- Note: the travis error does not seem to have occurred in this PR.

File: keras/applications/imagenet_utils.py [hunk 15]
```diff
@@ -18,26 +18,32 @@ def _preprocess_numpy_input(x, data_format, mode):
         x -= 1.
         return x
 
-    if data_format == 'channels_first':
-        if x.ndim == 3:
+    if mode == 'torch':
```
Comment (fchollet) [inline keras/applications/imagenet_utils.py]: I'm not sure there's anything torch-specific about this scheme (unlike the `caffe` scheme which was used by every pretrained model we had that came from Caffe). Or is it really torch-specific?

File: keras/applications/imagenet_utils.py [hunk 16]
```diff
@@ -18,26 +18,32 @@ def _preprocess_numpy_input(x, data_format, mode):
...
-        x[..., 2] -= 123.68
+            # 'RGB'->'BGR'
+            x = x[..., ::-1]
+        mean = [103.939, 116.779, 123.68]
+        std = None
+
+    if data_format == 'channels_first':
+        mean = np.expand_dims(np.expand_dims(mean, axis=1), axis=2)
+        if std is not None:
+            std = np.expand_dims(np.expand_dims(std, axis=1), axis=2)
```
Comment (fchollet) [inline keras/applications/imagenet_utils.py]: The problem of this approach is that it instantiates (and reshapes) a Numpy array at every call. Since this method is meant to be called in a loop, it will end up being expensive. it would be preferable to follow the previous pattern to avoid this.

File: keras/applications/imagenet_utils.py (hunk 15 above)
Comment (taehoonlee) [reply keras/applications/imagenet_utils.py]: To be precise, this is not torch-specific. This scheme is originated from [Torch ResNets](https://github.com/facebook/fb.resnet.torch/blob/master/pretrained/classify.lua) and many Torch implementations ([ResNeXt](https://github.com/facebookresearch/ResNeXt), [DenseNet](https://github.com/liuzhuang13/DenseNet), ...) follow the scheme.

File: keras/applications/imagenet_utils.py (hunk 16 above)
Comment (taehoonlee) [reply keras/applications/imagenet_utils.py]: I agree. Reverted.
